src/java.desktop/share/classes/sun/font/HBShaper.java

```diff
@@ -0,0 +1,643 @@
+        // so we need to keep a reference to it here until
+        // this FaceRef is collected.
+        private MemorySegment get_table_data_fn;
+
+        private FaceRef(Font2D font) {
+            this.font2D = font;
+        }
+
+        private synchronized MemorySegment getFace() {
+            if (face == null) {
+                createFace();
+                if (face != null) {
+                    Disposer.addObjectRecord(font2D, this);
+                }
+                font2D = null;
+            }
+            return face;
+        }
+
+        private void createFace() {
+            try {
+                get_table_data_fn = getBoundUpcallStub(Arena.ofAuto(),
+                        Font2D.class,
+                        font2D,                      // bind arg
+                        "getFontTableData",          // method name
+                        JAVA_INT,                   // return type
+                        JAVA_INT, ADDRESS); // arg types
+                if (get_table_data_fn == null) {
+                    return;
+                }
+                face = (MemorySegment)create_face_handle.invokeExact(get_table_data_fn);
+            } catch (Throwable t) {
+            }
+        }
+
+        @Override
+        public void dispose() {
+            try {
+                dispose_face_handle.invokeExact(face);
+            } catch (Throwable t) {
+            }
+        }
+    }
+
+
+    /* Upcall to receive results of layout */
+    private static void store_layout_results(
+        int slot,
+        int baseIndex,
+        int offset,
+        float startX,
+        float startY,
+        float devScale,
+        int charCount,
+        int glyphCount,
+        MemorySegment /* hb_glyph_info_t* */ glyphInfo,
+        MemorySegment /* hb_glyph_position_t* */ glyphPos
+        ) {
+
+        GVData gvdata = scopedVars.get().gvData();
+        Point2D.Float startPt = scopedVars.get().point();
+        float x=0, y=0;
+        float advX, advY;
+        float scale = 1.0f / HBFloatToFixedScale / devScale;
+
+        int initialCount = gvdata._count;
+
+        int maxGlyphs = (charCount > glyphCount) ? charCount : glyphCount;
+        int maxStore = maxGlyphs + initialCount;
+        boolean needToGrow = (maxStore > gvdata._glyphs.length) ||
+                             ((maxStore * 2 + 2) > gvdata._positions.length);
+        if (needToGrow) {
+            gvdata.grow(maxStore-initialCount);
+        }
+
+        int glyphPosLen = glyphCount * 2 + 2;
+        long posSize = glyphPosLen * PositionLayout.byteSize();
+        MemorySegment glyphPosArr = glyphPos.reinterpret(posSize);
+
+        long glyphInfoSize = glyphCount * GlyphInfoLayout.byteSize();
+        MemorySegment glyphInfoArr = glyphInfo.reinterpret(glyphInfoSize);
+
+         for (int i=0; i<glyphCount; i++) {
+             int storei = i + initialCount;
+             int cluster = (int)clusterHandle.get(glyphInfoArr, i) - offset;
```

Comment:
All the var handle calls in this loop are not exact - e.g. they use a `int` offset instead of a `long` one. Because of this, the memory access cannot be fully optimized. Adding a cast to `long` on all offsets coordinates yields a significant performance boost.

To avoid issues like these, it is recommended to set up the var handle using the `VarHandle::withInvokeExactBehavior` method, which will cause an exception to be thrown in case there's a type mismatch (similar to `MethodHandle::invokeExact`).